### PR TITLE
refactor(planner): apply constant folder at physical plan builder

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

- constant fold all expressions at `PhysicalPlanBuilder`
- rewrite all filter expr to `is_true(try_to_boolean(<expr>))` and remove all `value.as_non_null_boolean()` on evaluation.
- pushdown's filter is changed from `Vec<Expr>` to `Option<Expr>`

### CODE REVIEW

File: src/query/expression/src/types/number.rs [hunk 1]
```diff
@@ -549,24 +549,29 @@ impl<T: Number> SimpleDomain<T> {
...
     }
 }
 
-fn overflow_cast_with_minmax<T: Number, U: Number>(src: T, min: U, max: U) -> (U, bool) {
+fn overflow_cast_with_minmax<T: Number, U: Number>(
+    src: T,
+    min: U,
+    max: U,
+    fallback: U,
+) -> (U, bool) {
```
Comment (andylokandy) [inline src/query/expression/src/types/number.rs:567]:
```suggestion
fn overflow_cast_with_minmax<T: Number, U: Number>(
    src: T,
    min: U,
    max: U,
) -> Option<(U, bool)> {
```

Comment (sundy-li) [reply src/query/expression/src/types/number.rs:567]: For float to int, there's nan/inf case that we can't use unwrap.

**Comment by BohuTANG** [comment]
Conflicting files
src/query/storages/fuse/src/operations/read/native_data_source_deserializer.rs

File: src/query/expression/src/types/number.rs [hunk 2]
```diff
@@ -549,24 +549,29 @@ impl<T: Number> SimpleDomain<T> {
     }
 
     pub fn overflow_cast_with_minmax<U: Number>(&self, min: U, max: U) -> (SimpleDomain<U>, bool) {
-        let (min, min_overflowing) = overflow_cast_with_minmax::<T, U>(self.min, min, max);
-        let (max, max_overflowing) = overflow_cast_with_minmax::<T, U>(self.max, min, max);
+        let (min, min_overflowing) = overflow_cast_with_minmax::<T, U>(self.min, min, max, min);
+        let (max, max_overflowing) = overflow_cast_with_minmax::<T, U>(self.max, min, max, max);
```
Comment (andylokandy) [inline src/query/expression/src/types/number.rs:553]: fallback value is the best use case for Option, and we can use it like this:
```suggestion
        let (min, min_overflowing) = overflow_cast_with_minmax::<T, U>(self.min, min, max).unwrap_or(min);
        let (max, max_overflowing) = overflow_cast_with_minmax::<T, U>(self.max, min, max).unwrap_or(max);
```

File: src/query/expression/src/types/number.rs [hunk 3]
```diff
@@ -549,24 +549,29 @@ impl<T: Number> SimpleDomain<T> {
...
     let dest_min: T = num_traits::cast(min).unwrap_or(T::MIN);
     let dest_max: T = num_traits::cast(max).unwrap_or(T::MAX);
     let src_clamp: T = src.clamp(dest_min, dest_max);
     let overflowing = src != src_clamp;
-    // The number must be within the range that `U` can represent after clamping, therefore
-    // it's safe to unwrap.
-    let dest: U = num_traits::cast(src_clamp).unwrap();
 
+    // It will have errors if the src type is Inf/NaN
+    let dest: U = num_traits::cast(src_clamp).unwrap_or(fallback);
```
Comment (andylokandy) [inline src/query/expression/src/types/number.rs:574]:
```suggestion
    let dest: U = num_traits::cast(src_clamp)?;
```

File: src/query/functions/src/scalars/mod.rs [hunk 4]
```diff
@@ -77,6 +77,13 @@ fn register_auto_cast_rules(registry: &mut FunctionRegistry) {
     registry.register_default_cast_rules(CAST_FROM_VARIANT_RULES());
     registry.register_auto_try_cast_rules(CAST_FROM_VARIANT_RULES());
 
+    for data_type in ALL_INTEGER_TYPES {
+        registry.register_additional_cast_rules("is_true", vec![(
```
Comment (andylokandy) [inline src/query/functions/src/scalars/mod.rs:81]: Why do we need this rule?

File: src/query/storages/hive/hive/src/hive_table.rs [hunk 5]
```diff
@@ -247,7 +245,18 @@ impl HiveTable {
...
-            let columns = Self::get_columns_from_expressions(filters);
+            let columns = filter
+                .as_ref()
+                .map(|f| {
+                    let expr = f.as_expr(&BUILTIN_FUNCTIONS);
+                    let mut cols = HashSet::new();
+                    for (k, _) in expr.column_refs() {
+                        cols.insert(k);
+                    }
+                    cols
```
Comment (andylokandy) [inline src/query/storages/hive/hive/src/hive_table.rs:256]:
```suggestion
                    expr.column_refs().keys().collect::<HashSet<_>>()
```

File: tests/sqllogictests/suites/base/03_dml/03_0025_delete_from [hunk 6]
```diff
@@ -119,13 +119,13 @@ select count(*) = 0 from t
 statement ok
 insert into t values (1),(2),(NULL)
 
-statement ok
+statement error 1001
 delete from t where now()
```
Comment (andylokandy) [inline tests/sqllogictests/suites/base/03_dml/03_0025_delete_from:123]: Is this case related to the number-to-bool rule for `is_try`?

Comment (sundy-li) [reply tests/sqllogictests/suites/base/03_dml/03_0025_delete_from:123]: I will add is_deterministic check for deletion.

**Review by sundy-li — APPROVED** [review]
